dask/blockwise.py

```diff
@@ -708,6 +738,14 @@ def _cull_dependencies(self, all_hlg_keys, output_blocks):
                         deps.add(tups)
             key_deps[(self.output,) + out_coords] = deps | const_deps
 
+        # Add valid-key dependencies from io_deps
+        for key, io_dep in self.io_deps.items():
+            if io_dep.valid:
+                for out_coords in output_blocks:
+                    key = (self.output,) + out_coords
+                    valid_key_dep = io_dep[out_coords]
```

Comment:
This seems pretty specific to the `BlockwiseDep` written here: just mapping block indices to a key name.

What if the `BlockwiseDep` depends on some other keys in a more complex way, like producing tasks, or returning a scalar in some blocks and a dependency key in others?

Reply:
What I'm still puzzling through: in our `HighLevelGraph.from_collections`, we've declared the dependencies on `dfs` for this layer. Is there a way we can read those dependencies off of that, rather than assuming a structure here?

To answer my own question: I suppose those dependencies are unordered, and we would then be unable to cull as you've done here, mapping indices to delayed keys.

Reply:
You're getting the evolution of my thoughts here as I read more. So, if I understand correctly, `valid=True` implies exactly this structure for the `BlockwiseDep` values. If it returns anything other than a key, that would be an invalid `valid`, so to speak.